nixos/modules/services/hardware/thermald.nix

```diff
@@ -4,27 +4,392 @@ with lib;
+        };
+      };
+
+      WritePrefix = mkOption {
+        type = nullOr string;
+        default = null;
+        description = ''
+          Write some prefix attached to state value, like below the prefix is "level ". It will
+          preserve spaces as entered when writing to sysfs
+        '';
+      };
+    };
+  };
+
+  makeThermalConfigurationXml = configFile:
+    "<?xml version=\"1.0\"?>\n" +
+    xmlBlock "ThermalConfiguration" (map makePlatformXml configFile.platforms);
+
+  makePlatformXml = pCfg: xmlBlock "Platform" [
+    (xmlElement pCfg "Name")
+    (xmlElement pCfg "UUID")
+    (xmlElement pCfg "ProductName")
+    (xmlElement pCfg "Preference")
+    (xmlBlock "ThermalSensors" (map makeThermalSensorXml pCfg.ThermalSensors))
+    (xmlBlock "ThermalZones" (map makeThermalZoneXml pCfg.ThermalZones))
+    (xmlBlock "CoolingDevices" (map makeCoolingDeviceXml pCfg.CoolingDevices))
+  ];
+
+  makeThermalSensorXml = tsCfg: xmlBlock "ThermalSensor" [
+    (xmlElement tzCfg "Type")
+    (xmlElement tzCfg "Path")
+    (mapXmlElement boolToInt tzCfg "AsyncCapable")
+  ];
+
+  makeThermalZoneXml = tzCfg: xmlBlock "ThermalZone" [
+    (xmlElement tzCfg "Type")
+    (xmlElement tzCfg "Path")
+    (mapXmlElement boolToInt tzCfg "AsyncCapable")
+    (xmlBlock "TripPoints" (map makeTripPointXml tzCfg.TripPoints))
+  ];
+
+  makeTripPointXml = tpCfg: xmlBlock "TripPoint" ([
+    (xmlElement tpCfg "SensorType")
+    (mapXmlElement toString tpCfg "Temperature")
+    (xmlElement tpCfg "Type")
+    (xmlElement tpCfg "ControlType")
+  ] ++ (map makeTripPointCoolingDeviceXml tpCfg.CoolingDevices));
+
+  makeTripPointCoolingDeviceXml = tpcdCfg: xmlBlock "CoolingDevice" [
+    (mapXmlElement toString tpcdCfg "Index")
+    (xmlElement tpcdCfg "Type")
+    (mapXmlElement toString tpcdCfg "Influence")
+    (mapXmlElement toString tpcdCfg "SamplingPeriod")
+    (mapXmlElement toString tpcdCfg "TargetState")
+  ];
+
+  makeCoolingDeviceXml = cdCfg: xmlBlock "CoolingDevice" [
+    (xmlElement cdCfg "Type")
+    (mapXmlElement toString cdCfg "MinState")
+    (mapXmlElement toString cdCfg "IncDecStep")
+    (mapXmlElement toString cdCfg "ReadBack")
+    (mapXmlElement toString cdCfg "MaxState")
+    (mapXmlElement toString cdCfg "DebouncePeriod")
+    (xmlBlock "PidControl" [
+      (xmlElement cdCfg.PidControl "kp")
+      (xmlElement cdCfg.PidControl "kd")
+      (xmlElement cdCfg.PidControl "ki")
+    ])
+    (xmlElement cdCfg "WritePrefix")
+  ];
 in {
   ###### interface
-  options = { 
-    services.thermald = { 
+  options = {
+    services.thermald = {
       enable = mkOption {
         default = false;
         description = ''
           Whether to enable thermald, the temperature management daemon.
-        ''; 
-      };  
-    };  
-  };  
+        '';
+      };
+
+      debug = mkOption {
+        default = false;
+        description = ''
+          Whether to enable debug logging
+        '';
+      };
+
+      configFile = mkOption {
```

Comment:
I think it would be better to name this option `config` and introduce a new option `configFile`, which will default to the generated config, but will then be overridable by the user. Then `cfg.configFile` can also be passed directly to the exec script and to `etc`.